File: IOS/Packages/LiquorEngine/Sources/LiquorEngine/CollectionStats.swift
import Foundation

/// What the collection looks like, in one screen.
///
/// The research is blunt about why this exists: OnlyDrams' growth loop is a
/// **shareable stats screenshot**. Post titles tell the story — *"My OnlyDrams
/// Post (10 years of collecting)"*, *"Can some one do the OnlyDrams thing for
/// me?"* — and the pie chart is the marketing asset. *"Build something
/// screenshot-worthy."*
///
/// **Everything here counts the COLLECTION, never the drinking.** Bottles
/// owned, classes represented, distilleries, the oldest thing on the shelf.
/// Nothing counts pours, nothing is a streak, nothing goes up when you drink
/// more, and there is no total that rewards volume.
///
/// That is not only Apple guideline 1.4.3, which rejects apps encouraging
/// excessive consumption. It is also what the research found people actually
/// like: an app pitched on *"collection progress and competition"* posted three
/// times to r/bourbon and scored 1, 1, 1. What generates unprompted delight is
/// a bottle's own properties, not a scoreboard.
public enum CollectionStats: Sendable {

    /// One bottle, reduced to what a summary needs.
    public struct Entry: Hashable, Sendable {
        public let name: String?
        public let classType: ClassType?
        public let distillery: String?
        public let brand: String?
        public let abv: Double?
        public let isOpen: Bool
        public let isFinished: Bool
        /// A store pick, single barrel or anything else with barrel detail.
        public let isPick: Bool
        public let addedAt: Date
        public let openedAt: Date?
        public let ageMonths: Int?
        public let storageLocation: String?
        public let purchasePriceCents: Int?
        public let purchasedAt: Date?
        public let costPerPourCents: Int?

        public init(
            name: String? = nil,
            classType: ClassType? = nil,
            distillery: String? = nil,
            brand: String? = nil,
            abv: Double? = nil,
            isOpen: Bool = false,
            isFinished: Bool = false,
            isPick: Bool = false,
            addedAt: Date = Date(),
            openedAt: Date? = nil,
            ageMonths: Int? = nil,
            storageLocation: String? = nil,
            purchasePriceCents: Int? = nil,
            purchasedAt: Date? = nil,
            costPerPourCents: Int? = nil
        ) {
            self.name = name
            self.classType = classType
            self.distillery = distillery
            self.brand = brand
            self.abv = abv
            self.isOpen = isOpen
            self.isFinished = isFinished
            self.isPick = isPick
            self.addedAt = addedAt
            self.openedAt = openedAt
            self.ageMonths = ageMonths
            self.storageLocation = storageLocation
            self.purchasePriceCents = purchasePriceCents
            self.purchasedAt = purchasedAt
            self.costPerPourCents = costPerPourCents
        }
    }

    /// Money for one label -- a year, usually. Kept apart from `Slice`
    /// because cents are not a count and a chart must not treat them as one.
    public struct Amount: Hashable, Sendable, Identifiable {
        public let label: String
        public let cents: Int
        public var id: String { label }
    }

    /// One named bottle and a number about it, for the notable rows.
    public struct Standout: Hashable, Sendable {
        public let name: String
        public let value: Int
    }

    /// A slice of the collection, for a chart or a list.
    public struct Slice: Hashable, Sendable, Identifiable {
        public let label: String
        public let count: Int
        public var id: String { label }

        public func share(of total: Int) -> Double {
            total > 0 ? Double(count) / Double(total) : 0
        }
    }

    public struct Summary: Sendable, Equatable {
        /// On the shelf. Finished bottles are counted separately and never
        /// folded in: a collection is what you have, not what you have had.
        public let onShelf: Int
        public let open: Int
        public let sealed: Int
        /// Archived, not deleted. Reported as a plain number with no
        /// celebration attached.
        public let finished: Int

        public let byClass: [Slice]
        public let byDistillery: [Slice]
        public let byStrength: [Slice]

        /// Brands, so "how much Weller do I have" is one line.
        public let byBrand: [Slice]
        /// Where bottles are kept. Empty until somebody records a place.
        public let byPlace: [Slice]
        /// Bottles added per calendar year, oldest year first. Growth of the
        /// COLLECTION, never a count of anything drunk.
        public let addedByYear: [Slice]

        public let distilleryCount: Int
        public let classCount: Int
        /// Store picks and single barrels on the shelf -- the bottles this
        /// app exists for.
        public let picks: Int
        /// The strongest thing on the shelf, as a property of a bottle rather
        /// than an achievement.
        public let highestProof: Double?
        public let oldestStatedAgeMonths: Int?
        /// The open bottle that has been open longest, in days. A fact about
        /// oxidation, not a prompt to finish it.
        public let longestOpen: Standout?

        // Money. The screen shows none of this unless the person turned the
        // shelf-value switch on; the research is clear that the number is
        // wanted by some and actively avoided by others.
        /// Spend per calendar year of purchase, oldest first. Bottles with a
        /// price but no purchase date are left out rather than guessed.
        public let spentByYear: [Amount]
        /// Mean cost of a pour across the priced bottles on the shelf.
        public let averageCostPerPourCents: Int?
        /// The most and least expensive pours on the shelf.
        public let dearestPour: Standout?
        public let cheapestPour: Standout?

        public var isEmpty: Bool { onShelf == 0 && finished == 0 }
    }

    /// Strength bands, in the words people use rather than numbers.
    ///
    /// Bands not a histogram: the interesting fact is "I own a lot of
    /// barrel-proof bourbon", and a bar chart of exact ABVs says that less
    /// clearly than four labels do.
    static func strengthBand(_ abv: Double) -> String {
        switch abv {
        case ..<45: return "80–89 proof"
        case ..<50: return "90–99 proof"
        case ..<57.5: return "100–114 proof"
        default: return "115 proof and up"
        }
    }

    public static func summarise(
        _ entries: [Entry],
        now: Date = Date(),
        calendar: Calendar = .current
    ) -> Summary {
        let live = entries.filter { !$0.isFinished }
        let finished = entries.count - live.count

        let byClass = tally(live.compactMap { $0.classType?.label })
        let byDistillery = tally(live.compactMap(\.distillery))
        let byStrength = tally(live.compactMap { $0.abv.map(strengthBand) })
        let byBrand = tally(live.compactMap(\.brand))
        let byPlace = tally(live.compactMap {
            $0.storageLocation?.trimmingCharacters(in: .whitespaces)
        }.filter { !$0.isEmpty })

        // Years run oldest first: a growth line reads left to right. Every
        // bottle counts here, finished included -- it was added that year.
        let addedByYear = tally(entries.map { String(calendar.component(.year, from: $0.addedAt)) })
            .sorted { $0.label < $1.label }

        let longestOpen = live
            .filter(\.isOpen)
            .compactMap { entry -> Standout? in
                guard let name = entry.name,
                      let days = AgeMath.daysOpen(openedAt: entry.openedAt, now: now, calendar: calendar)
                else { return nil }
                return Standout(name: name, value: days)
            }
            .max { $0.value < $1.value }

        // Priced live bottles with a purchase date, grouped by that year.
        var spent: [String: Int] = [:]
        for entry in live {
            guard let cents = entry.purchasePriceCents, let at = entry.purchasedAt else { continue }
            spent[String(calendar.component(.year, from: at)), default: 0] += cents
        }
        let spentByYear = spent
            .map { Amount(label: $0.key, cents: $0.value) }
            .sorted { $0.label < $1.label }

        let pours = live.compactMap { entry -> Standout? in
            guard let name = entry.name, let cents = entry.costPerPourCents else { return nil }
            return Standout(name: name, value: cents)
        }
        let averagePour = pours.isEmpty
            ? nil
            : Int((Double(pours.reduce(0) { $0 + $1.value }) / Double(pours.count)).rounded())

        return Summary(
            onShelf: live.count,
            open: live.filter(\.isOpen).count,
            sealed: live.filter { !$0.isOpen }.count,
            finished: finished,
            byClass: byClass,
            byDistillery: byDistillery,
            byStrength: byStrength,
            byBrand: byBrand,
            byPlace: byPlace,
            addedByYear: addedByYear,
            distilleryCount: Set(live.compactMap(\.distillery)).count,
            classCount: Set(live.compactMap { $0.classType?.label }).count,
            picks: live.filter(\.isPick).count,
            highestProof: live.compactMap(\.abv).max().map { ABV(percent: $0).proof },
            oldestStatedAgeMonths: live.compactMap(\.ageMonths).max(),
            longestOpen: longestOpen,
            spentByYear: spentByYear,
            averageCostPerPourCents: averagePour,
            dearestPour: pours.max { $0.value < $1.value },
            cheapestPour: pours.min { $0.value < $1.value })
    }

    /// Biggest first, then alphabetically so equal counts do not reshuffle
    /// between launches — a chart whose order changes on its own reads as a
    /// bug.
    static func tally(_ values: [String]) -> [Slice] {
        var counts: [String: Int] = [:]
        for value in values { counts[value, default: 0] += 1 }
        return counts
            .map { Slice(label: $0.key, count: $0.value) }
            .sorted {
                $0.count != $1.count
                    ? $0.count > $1.count
                    : $0.label.localizedCaseInsensitiveCompare($1.label) == .orderedAscending
            }
    }
}

/// What is open right now, written out for somebody else to read.
///
/// From the research, unprompted: *"I really like the menu concept! Now if
/// there was a way to take the spreadsheet and populate the menu......"*
///
/// It is a list of what is OPEN, because that is the question a guest is
/// actually asking. A sealed bottle is not on offer, and printing the whole
/// collection turns a menu into a brag.
public enum PourMenu: Sendable {

    public struct Item: Hashable, Sendable {
        public let name: String
        public let detail: String?
        public let proof: Double?

        public init(name: String, detail: String? = nil, proof: Double? = nil) {
            self.name = name
            self.detail = detail
            self.proof = proof
        }
    }

    /// Plain text, because it has to survive being pasted into a message.
    ///
    /// No prices. A menu with prices on it reads as bragging about what the
    /// evening cost, and the one thing a guest cannot do with that information
    /// is enjoy the whiskey.
    public static func text(title: String, items: [Item]) -> String {
        guard !items.isEmpty else {
            return "\(title)\n\nNothing open at the moment."
        }
        let lines = items.map { item -> String in
            var line = "• \(item.name)"
            if let proof = item.proof {
                line += String(format: "  (%.1f proof)", proof)
            }
            if let detail = item.detail, !detail.isEmpty {
                line += "\n  \(detail)"
            }
            return line
        }
        return ([title, ""] + lines).joined(separator: "\n")
    }
}

File: Android/engine/src/main/kotlin/com/talastack/liquorlog/engine/CollectionStats.kt
package com.talastack.liquorlog.engine

import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.util.Locale
import kotlin.math.roundToInt

/**
 * What the collection looks like, in one screen.
 *
 * The research is blunt about why this exists: OnlyDrams' growth loop is a
 * **shareable stats screenshot**. Post titles tell the story -- *"My OnlyDrams
 * Post (10 years of collecting)"*, *"Can some one do the OnlyDrams thing for
 * me?"* -- and the pie chart is the marketing asset. *"Build something
 * screenshot-worthy."*
 *
 * **Everything here counts the COLLECTION, never the drinking.** Bottles
 * owned, classes represented, distilleries, the oldest thing on the shelf.
 * Nothing counts pours, nothing is a streak, nothing goes up when you drink
 * more, and there is no total that rewards volume.
 *
 * That is not only Apple guideline 1.4.3, which rejects apps encouraging
 * excessive consumption. It is also what the research found people actually
 * like: an app pitched on *"collection progress and competition"* posted three
 * times to r/bourbon and scored 1, 1, 1. What generates unprompted delight is
 * a bottle's own properties, not a scoreboard.
 */
object CollectionStats {

    /** One bottle, reduced to what a summary needs. */
    data class Entry(
        val name: String? = null,
        val classType: ClassType? = null,
        val distillery: String? = null,
        val brand: String? = null,
        val abv: Double? = null,
        val isOpen: Boolean = false,
        val isFinished: Boolean = false,
        /**
         * A sample rather than a bottle. Counted on its own line and kept out
         * of the bottle counts and breakdowns: fifty millilitres from a friend
         * is not a bottle of Weller on the shelf.
         */
        val isSample: Boolean = false,
        /** A store pick, single barrel or anything else with barrel detail. */
        val isPick: Boolean = false,
        val addedAt: Instant = Instant.now(),
        val openedAt: Instant? = null,
        val ageMonths: Int? = null,
        val storageLocation: String? = null,
        val purchasePriceCents: Int? = null,
        val purchasedAt: Instant? = null,
        val costPerPourCents: Int? = null,
        /** The Blanton's stopper letter, when the bottle has one. */
        val topperLetter: String? = null
    )

    /**
     * Money for one label -- a year, usually. Kept apart from [Slice] because
     * cents are not a count and a chart must not treat them as one.
     */
    data class Amount(val label: String, val cents: Int) {
        val id: String get() = label
    }

    /** One named bottle and a number about it, for the notable rows. */
    data class Standout(val name: String, val value: Int)

    /** A slice of the collection, for a chart or a list. */
    data class Slice(val label: String, val count: Int) {
        val id: String get() = label

        fun share(of: Int): Double = if (of > 0) count.toDouble() / of.toDouble() else 0.0
    }

    data class Summary(
        /**
         * On the shelf. Finished bottles are counted separately and never
         * folded in: a collection is what you have, not what you have had.
         */
        val onShelf: Int,
        val open: Int,
        val sealed: Int,
        /**
         * Archived, not deleted. Reported as a plain number with no
         * celebration attached.
         */
        val finished: Int,
        /** Samples on hand, not finished. Never folded into [onShelf]. */
        val samples: Int,
        val byClass: List<Slice>,
        val byDistillery: List<Slice>,
        val byStrength: List<Slice>,
        /** Brands, so "how much Weller do I have" is one line. */
        val byBrand: List<Slice>,
        /** Where bottles are kept. Empty until somebody records a place. */
        val byPlace: List<Slice>,
        /**
         * Bottles added per calendar year, oldest year first. Growth of the
         * COLLECTION, never a count of anything drunk.
         */
        val addedByYear: List<Slice>,
        val distilleryCount: Int,
        val classCount: Int,
        /**
         * Store picks and single barrels on the shelf -- the bottles this app
         * exists for.
         */
        val picks: Int,
        /**
         * The strongest thing on the shelf, as a property of a bottle rather
         * than an achievement.
         */
        val highestProof: Double?,
        val oldestStatedAgeMonths: Int?,
        /**
         * The open bottle that has been open longest, in days. A fact about
         * oxidation, not a prompt to finish it.
         */
        val longestOpen: Standout?,
        /**
         * Which Blanton's stopper letters are on the shelf. Null until there
         * is at least one, so a shelf with no Blanton's never sees the word.
         */
        val topperLetters: TopperLetters.Progress?,
        /**
         * Spend per calendar year of purchase, oldest first. Bottles with a
         * price but no purchase date are left out rather than guessed.
         */
        val spentByYear: List<Amount>,
        /** Mean cost of a pour across the priced bottles on the shelf. */
        val averageCostPerPourCents: Int?,
        /** The most and least expensive pours on the shelf. */
        val dearestPour: Standout?,
        val cheapestPour: Standout?
    ) {
        val isEmpty: Boolean get() = onShelf == 0 && finished == 0 && samples == 0
    }

    /**
     * Strength bands, in the words people use rather than numbers.
     *
     * Bands not a histogram: the interesting fact is "I own a lot of
     * barrel-proof bourbon", and a bar chart of exact ABVs says that less
     * clearly than four labels do.
     */
    internal fun strengthBand(abv: Double): String = when {
        abv < 45 -> "80–89 proof"
        abv < 50 -> "90–99 proof"
        abv < 57.5 -> "100–114 proof"
        else -> "115 proof and up"
    }

    fun summarise(
        entries: List<Entry>,
        now: Instant = Instant.now(),
        zone: ZoneId = ZoneId.systemDefault()
    ): Summary {
        val unfinished = entries.filter { !it.isFinished }
        val finished = entries.size - unfinished.size
        val samples = unfinished.count { it.isSample }
        val live = unfinished.filter { !it.isSample }

        val byClass = tally(live.mapNotNull { it.classType?.label })
        val byDistillery = tally(live.mapNotNull { it.distillery })
        val byStrength = tally(live.mapNotNull { it.abv?.let { abv -> strengthBand(abv) } })
        val byBrand = tally(live.mapNotNull { it.brand })
        val byPlace = tally(
            live.mapNotNull { it.storageLocation?.trimSpaces() }.filter { it.isNotEmpty() }
        )

        // Years run oldest first: a growth line reads left to right. Every
        // bottle counts here, finished included -- it was added that year.
        val addedByYear = tally(entries.map { year(it.addedAt, zone) })
            .sortedBy { it.label }

        val longestOpen = live
            .filter { it.isOpen }
            .mapNotNull { entry ->
                val name = entry.name ?: return@mapNotNull null
                val days = AgeMath.daysOpen(entry.openedAt, now) ?: return@mapNotNull null
                Standout(name, days)
            }
            .maxByOrNull { it.value }

        // Priced live bottles with a purchase date, grouped by that year.
        val spent = mutableMapOf<String, Int>()
        for (entry in live) {
            val cents = entry.purchasePriceCents ?: continue
            val at = entry.purchasedAt ?: continue
            val label = year(at, zone)
            spent[label] = (spent[label] ?: 0) + cents
        }
        val spentByYear = spent
            .map { (label, cents) -> Amount(label, cents) }
            .sortedBy { it.label }

        val letters = live.mapNotNull { it.topperLetter }
        val toppers = if (letters.isEmpty()) null else TopperLetters.progress(letters)

        val pours = live.mapNotNull { entry ->
            val name = entry.name ?: return@mapNotNull null
            val cents = entry.costPerPourCents ?: return@mapNotNull null
            Standout(name, cents)
        }
        val averagePour = if (pours.isEmpty()) {
            null
        } else {
            (pours.sumOf { it.value }.toDouble() / pours.size).roundToInt()
        }

        return Summary(
            onShelf = live.size,
            open = live.count { it.isOpen },
            sealed = live.count { !it.isOpen },
            finished = finished,
            samples = samples,
            byClass = byClass,
            byDistillery = byDistillery,
            byStrength = byStrength,
            byBrand = byBrand,
            byPlace = byPlace,
            addedByYear = addedByYear,
            distilleryCount = live.mapNotNull { it.distillery }.toSet().size,
            classCount = live.mapNotNull { it.classType?.label }.toSet().size,
            picks = live.count { it.isPick },
            highestProof = live.mapNotNull { it.abv }.maxOrNull()?.let { ABV(it).proof },
            oldestStatedAgeMonths = live.mapNotNull { it.ageMonths }.maxOrNull(),
            longestOpen = longestOpen,
            topperLetters = toppers,
            spentByYear = spentByYear,
            averageCostPerPourCents = averagePour,
            dearestPour = pours.maxByOrNull { it.value },
            cheapestPour = pours.minByOrNull { it.value }
        )
    }

    /**
     * Biggest first, then alphabetically so equal counts do not reshuffle
     * between launches -- a chart whose order changes on its own reads as a
     * bug.
     */
    internal fun tally(values: List<String>): List<Slice> {
        val counts = LinkedHashMap<String, Int>()
        for (value in values) counts[value] = (counts[value] ?: 0) + 1
        return counts
            .map { (label, count) -> Slice(label, count) }
            .sortedWith(
                compareByDescending<Slice> { it.count }
                    .thenBy(String.CASE_INSENSITIVE_ORDER) { it.label }
            )
    }

    private fun year(at: Instant, zone: ZoneId): String =
        LocalDate.ofInstant(at, zone).year.toString()

    /** Spaces and tabs, as Swift's `.whitespaces` trims; not line breaks. */
    private fun String.trimSpaces(): String =
        trim { it.isWhitespace() && it != '\n' && it != '\r' }
}

/**
 * What is open right now, written out for somebody else to read.
 *
 * From the research, unprompted: *"I really like the menu concept! Now if
 * there was a way to take the spreadsheet and populate the menu......"*
 *
 * It is a list of what is OPEN, because that is the question a guest is
 * actually asking. A sealed bottle is not on offer, and printing the whole
 * collection turns a menu into a brag.
 */
object PourMenu {

    data class Item(
        val name: String,
        val detail: String? = null,
        val proof: Double? = null
    )

    /**
     * Plain text, because it has to survive being pasted into a message.
     *
     * No prices. A menu with prices on it reads as bragging about what the
     * evening cost, and the one thing a guest cannot do with that information
     * is enjoy the whiskey.
     */
    fun text(title: String, items: List<Item>): String {
        if (items.isEmpty()) return "$title\n\nNothing open at the moment."
        val lines = items.map { item ->
            val line = StringBuilder("• ").append(item.name)
            item.proof?.let { line.append(String.format(Locale.ROOT, "  (%.1f proof)", it)) }
            val detail = item.detail
            if (!detail.isNullOrEmpty()) line.append("\n  ").append(detail)
            line.toString()
        }
        return (listOf(title, "") + lines).joinToString("\n")
    }
}
